### PR TITLE
Travis tests for daemon, fixes for SQLA sessions et al.

### LINKED ISSUE #464: Problem in Celery with SQLA
The daemon crashes at the first connection with SQLA.
Some random comments:
- We cannot upgrade to celery 4 because SQLA is not valid anymore as a broker
- We are using SQLA as a broker currently
- It is possible that there is a conflict between the SQLA connections of Celery and ours?
- We should go away from SQLA as a broker. Two ways: 1. we upgrade, and use RabbitMQ (note that however this requires a new service running on users' computers); 2. We stop using Celery - in the end, we don't really need it, we just need to run four functions every few seconds.
- Upgrading to the most recent 3.x version of celery does not solve the issue, see giovannipizzi/aiida_core@454db10c6d

### PR DESCRIPTION
**Pull request by giovannipizzi**
This PR adds a set of tests for the daemon (30 very short calculations are submitted and retrieval is checked). This is now implemented for both Django and SQLA, after the more 'classic' tests.

Also, a revisiting of how sessions are managed in SQLA has been done, to fix an issue that appeared while introducing the tests. Beside using a scoped_session (that should make sessions thread-safe according to SQLA docs, but this is still untested) we also made the global session fork-safe by registering a function that disposes the engine and also recreates a new sessionmaker (the latter is also essential!)

Note that however having a global session (rather than an object that is passed around) is *discouraged* by the SQLAlchemy docs, so we need to rethink this...

Also:
- upped dependencies for celery
- changed the way the column 'state' of DbNode (for SQLA) is managed (see changes in models/node, and in the dummy model, as well as a required change in the querybuilder_sqla)
- fixed the get_state() method of DbNode (SQLA) because it was caching the results (it was using internally DbNode.dbstates, the SQLA relationship, that however caches the results. Now it instead performs a new query when asked for the current state.

This PR fixes #464

### CODE REVIEW

File: aiida/orm/implementation/sqlalchemy/code.py [hunk 1]
```diff
@@ -130,7 +130,8 @@ def delete_code(code):
             "has {} output links".format(len(existing_outputs)))
     else:
         repo_folder = code._repository_folder
-        from aiida.backends.sqlalchemy import session
+        from aiida.backends.sqlalchemy import get_scoped_session
+        session = get_scoped_session()
         session.begin(subtransactions=True)
         try:
             code.dbnode.delete()
```
Comment (lekah) [inline aiida/orm/implementation/sqlalchemy/code.py]: This has nothing to do with any issues now, but shouldn't the order of folder.erase and session.commit be reversed?
I.e. only erase a folder if the commit went through?

**Comment by szoupanos** [comment]
For completeness, I also add a like to the SQLA scoped session documentation
http://docs.sqlalchemy.org/en/latest/orm/contextual.html

**Comment by szoupanos** [comment]
I had a fast look at the changes and I am a bit confused with the scoped session.
Maybe it is better if we discuss face-to-face.

In general my understanding is that when we import & use the session like:

```
import aiida.backends.sqlalchemy
aiida.backends.sqlalchemy.session.do_something()

```

we refer to the same session. At least within the same thread.
From http://docs.sqlalchemy.org/en/latest/orm/contextual.html I understand that this is what scopedsession does but in a different way (by calling `Session()`):

```
>>> from sqlalchemy.orm import scoped_session
>>> from sqlalchemy.orm import sessionmaker

>>> session_factory = sessionmaker(bind=some_engine)
>>> Session = scoped_session(session_factory)
>>> some_session = Session()
>>> some_other_session = Session()
>>> some_session is some_other_session
True
```

What remains unclear to me, is 
- why the approach with scopedsession works better (maybe the import approach was not implemented everywhere & also has its drawbacks) 
- and how the daemon problem was solved

**Comment by szoupanos** [comment]
Moreover, is there a test that demonstrates the problem with the state attribute that we saw yesterday (and it is fixed with this PR)?

File: aiida/orm/implementation/sqlalchemy/code.py (hunk 1 above)
Comment (giovannipizzi) [inline aiida/orm/implementation/sqlalchemy/code.py]: Du hast Recht! I fixed this.

**Comment by giovannipizzi** [comment]
The scoped session as it is implemented now should make sessions thread-safe (or if you want, you get always the same session if you are in the same thread, but different sessions in different threads).
Quoting the SQLAlchemy docs from the link above:

>  The scoped_session object by default uses this object as storage, so that a single Session is maintained for all who call upon the scoped_session registry, but only within the scope of a single thread. Callers who call upon the registry in a different thread get a Session instance that is local to that other thread.

On top of this, the real problem was with forking. The current logic implemented in `reset_session` as well as the content of the function registered with `register_after_fork` do the right thing to create new engines and sessions for new forks.

For what concerns the tests, the problem of the state caching was exposed by the daemon tests in this PR, and this is the only simple way (currently) to test this, as to show the effect we need to independent processes changing the content of the DB at the same time.

**Pull request merged**